internal/rules/format.go

```diff
@@ -0,0 +1,165 @@
+func (b *formatBuffer) writeIndented(indentFirst bool, fn func(*formatBuffer)) {
+	// Pass a temporary buffer to the closure to capture the written bytes.
+	var buf formatBuffer
+	fn(&buf)
+
+	// Indent the captured bytes and write them to the underlying
+	// strings.Builder.
+	writeIndented(&b.Builder, buf.String(), 2, indentFirst)
+}
+
+// writeIndentedList iterates the list of results and invokes fn for each
+// result, passing a *formatBuffer which will indent every line written to it
+// by 2 spaces.
+func (b *formatBuffer) writeIndentedList(results []EvalResult, fn func(*formatBuffer, *EvalResult)) {
+	b.writeIndented(true, func(buf *formatBuffer) {
+		for i, result := range results {
+			buf.WriteRune('\n')
+			fmt.Fprintf(buf, "%d) ", i+1)
+			buf.writeIndented(false, func(buf *formatBuffer) {
+				fn(buf, &result)
+			})
+		}
+	})
+}
+
+// formatFailure formats a failed result and writes the human readable
+// representation to buf. The caller must ensure that result.Success is false.
+func formatFailure(buf *formatBuffer, result *EvalResult) {
+	result = result.flatten()
+
+	formatRuleKind(buf, result.Rule.Kind())
+	formatRuleMeta(buf, result.Rule.Meta())
+
+	successes, failures := result.partitionNested()
+
+	switch r := result.Rule.(type) {
+	case *MatchRule:
+		fmt.Fprintf(buf, "Expected trust anchor %q was not found.\n", r.trustAnchor)
+	case *MatchRegexRule:
+		fmt.Fprintf(buf, "Trust anchor matching regular expression %q was not found.\n", r.pattern.String())
+	case *NotRule:
+		buf.WriteString("Expected nested rule to fail, but it did not:\n")
+		buf.writeIndentedList(successes, formatUnexpectedSuccess)
+	case *AllOfRule:
+		buf.WriteString("Expected ALL of the nested rules to match, but found ")
+
+		if len(failures) == 1 {
+			buf.WriteString("one failure:\n")
+		} else {
+			fmt.Fprintf(buf, "%d failures:\n", len(failures))
+		}
+
+		buf.writeIndentedList(failures, formatFailure)
+	case *AnyOfRule:
+		buf.WriteString("Expected ANY of the nested rule to match, but none did:\n")
+		buf.writeIndentedList(failures, formatFailure)
+	case *OneOfRule:
+		buf.WriteString("Expected EXACTLY ONE nested rule to match, but ")
+
+		if len(successes) == 0 {
+			buf.WriteString("none did:\n")
+			buf.writeIndentedList(failures, formatFailure)
+		} else {
+			fmt.Fprintf(buf, "found %d:\n", len(successes))
+			buf.writeIndentedList(successes, formatUnexpectedSuccess)
+		}
+	}
+}
+
+// formatUnexpectedSuccess format a result that was expected to fail, but
+// succeeded unexpectedly, and writes the human readable representation to buf.
+// The caller must ensure that result.Success is true.
+func formatUnexpectedSuccess(buf *formatBuffer, result *EvalResult) {
+	result = result.flatten()
+
+	formatRuleKind(buf, result.Rule.Kind())
+	formatRuleMeta(buf, result.Rule.Meta())
+
+	trustAnchors := result.Matched.Slice()
+	sort.Strings(trustAnchors)
+
+	buf.WriteString("Matched trust anchors:\n")
+	formatTrustAnchors(buf, result.Matched)
+}
+
```

Comment:
I was thinking more playing with the result status of the rule maybe. So in the first case we would have message like:
`Expected len(rules) to to not match, evaluated len(unmatched) and len(matched)` and just print the offending match rules.

Or

`Expected only one rule to match, evaluated len(matched)`  and again print the rules that matched.
or do you think something like that won't work?

Reply:
The implementation of `formatUnexpectedSuccess` isn't 100% like I want it because it's tricky to present this to the user in a more elegant way. It can be done, but require more work. There a reason why I'm not further descending into `result.Nested` here like i do in `formatFailure`: You have to negate every result along the way and tell the user the exact opposite of what happened in order to be helpful.

That's why I stopped here for now and only print out the matched trust anchors that where unexpected.

I'd like to leave this for a future PR that concentrate on further improving the error reporting here. We'll probably get a better idea of what can be improved here once we have more testdata and real-world use cases to play with.